# Practical 1

In this practical we will set up a working Python3 development environment and will start familiarizing a bit with Python.

## Slides

Slides of the introduction can be found here: [Intro](docs/sciprolabintro.pdf)

## Setting up the environment

We will need to install several pieces of software to get a working programming environment suitable for this practical. In this section we will install everything that we will need in the next few weeks. 

Python3 is available for Windows, Linux and Mac.

### Our toolbox

Basically we will need to install some software to start off with. If you work on Windows or Mac, skip the following and go straight to the section "Windows/Mac installation"

#### Linux installation 

1. The Python interpreter. In this course we will use python version 3.x. A lot of information on python can be found on the [python web page](https://www.python.org/). Open a terminal and try typing in:

    ```python3```
    
    if you get an error like "python3 command not found" you need to install it, while if you get something like this:
    
    ![](img/console.png)
    
    you are already sorted, just type Ctrl-D to exit.

    Installation on a debian-like linux distribution (e.g. Ubuntu) can be done by typing the following commands on a terminal:

    ```sudo apt-get update```
    
    ```sudo apt-get install python3```



2. Install now the package manager pip, which is a very convenient tool to install python packages, with the following command:
    
    ```sudo apt-get install python3-pip```
    
    <div class="alert alert-warning">

    **Note:** 
    
    If pip is already installed in your system you will get a message like: 
    python3-pip is already the newest version (3.x.y)

    </div>
    
3. Finally, install the Integrated Development Environment (IDE) that we will be using. This is called Visual Studio Code and is available for all platforms. You can read about it [here](https://code.visualstudio.com/). Downloads for all platforms can be found [here](https://code.visualstudio.com/Download).

#### Windows/Mac installation 

1. The python interpreter. In this course we will use python version 3.x. A lot of information on python can be found on the [python web page](https://www.python.org/). Installers for Windows and Mac can be found at [this page](https://www.python.org/). Click on Download Python 3.6.x. When executing the installer, please remember to flag "Add Python 3.6.2 to PATH" and then click on Install now (see picture below). 

    ![](img/installnow.png)


2. Install now the Integrated Development Environment (IDE) that we will be using. This is called Visual Studio Code and is available for all platforms. You can read about it [here](https://code.visualstudio.com/). Downloads for all platforms can be found [here](https://code.visualstudio.com/Download).


## The console

To access the console on Linux just open a terminal and type:

```python3```

while in Windows you have to look for "Python" and run "Python 3.6".
The console should look like this:

   ![](img/pythonconsole.png)
   

Now we are all set to start interacting with the Python interpreter. Type in the following:

In [1]:
5 + 3

8

All as expected. The "In [1]" line is the input, while the "Out [1]" reports the output of the interpreter. Let's challenge python with some other operations:

In [2]:
12 / 5

2.4

In [3]:
1/133

0.007518796992481203

And some assignments:

In [4]:
a = 10
b = 7
s = a + b
d = a / b

print("sum is:",s, " division is:",d)

sum is: 17  division is: 1.4285714285714286


In the first four lines, values have been assigned to variables through the = operator. In the last line, the print function is used to display the output. For the time being, we will skip all the details and just notice that the print function somehow managed to get text and variables in input and coherently merged them in an output text. 
Although quite useful in some occasions, the console is quite limited therefore you can close it for now. To exit type Ctrl-D or exit().

## Visual Studio Code

Once you open the IDE Visual Studio Code you will see the welcome screen:

![](img/visual_studio_code.png)

You can find useful information on this tool [here](https://code.visualstudio.com/docs#vscode). Please spend some time having a look at that page.

Once you are done with it you can close this window. First thing to do is to set the python interpreter to use. Click on View --> Command Palette and type "Python" in the text search space. Select **Python: Select Workspace Interpreter** as shown in the picture below.

![](img/pythoninterpreter.png)

Finally, select the python version you want to use (e.g. Python3).

Now you can click on **Open Folder** to create a new folder to place all the scripts you are going to create. You can call it something like "EXERCISES". Next you can create a new file, *example1.py* (*.py* extension stands for python). 

Visual Studio Code will understand that you are writing Python code and will help you with valid syntax for your program.

<div class="alert alert-warning">

**Warning:** 

If you get the following error message:

![](img/pylint.png)

click on **Install Pylint** which is a useful tool to help your coding experience.
</div>


Add the following text to your *example1.py* file.

In [5]:
"""
This is the first example of Python script.
"""
a = 10 # variable a
b = 33 # variable b
c = a / b # variable c holds the ratio

# Let's print the result to screen.
print("a:", a, " b:", b, " a/b=", c)

a: 10  b: 33  a/b= 0.30303030303030304


A couple of things worth nothing. The first three lines opened and closed by """ are some text describing the content of the script. Moreover, comments are proceeded by the hash key (#) and they are just ignored by the python interpreter. Please remember to comment your code, as it helps readability and will make your life easier when you have to modify or just understand the code you wrote some time in the past.  

Please notice that Visual Studio Code will help you writing your Python scripts. For example, when you start writing the **print** line it will complete the code for you (**if the Pylint extension mentioned above is installed**), suggesting the functions that match the letters written. This useful feature is called *code completion* and, alongside suggesting possible matches, it also visualizes a description of the function and parameters it needs. Here is an example:

![](img/codecompletion.png)

Save the file (Ctrl+S as shortcut). By right-clicking in the code panel and selecting **Run Python File in Terminal** (see picture below) you can execute the code you have just written. 

![](img/pythonrun.png)

Upon clicking on *Run Python File in Terminal* a terminal panel should pop up in the lower section of the coding panel and the result shown above should be reported.

Saving script files like the *example1.py* above is also handy because they can be invoked several times (later on we will learn how to get inputs from the command line to make them more useful...). To do so, you just need to call the python intepreter passing the script file as parameter. From the folder containing the *example1.py* script:

```python3 example1.py```

will in fact return:

a: 10  b: 33  a/b= 0.30303030303030304


*Please take some time to familiarize with Visual Studio Code (creating files, saving files etc.) as in the next practicals we will take this ability for granted.*

## The debugger

Another important feature of advanced Integrated Development Environments (IDEs) is their debugging capabilities. Visual Studio Code comes with a debugging tool that can help you trace the execution of your code and understand where possible errors hide. 

Write the following code on a new file (let's call it *integer_sum.py*) and execute it to get the result.

In [6]:
""" integer_sum.py is a script to
 compute the sum of the first 1200 integers. """

S = 0
for i in range(0, 1201):
    S = S + i

print("The sum of the first 1200 integers is: ", S)

The sum of the first 1200 integers is:  720600


Without getting into too many details, the code you just wrote starts initializing a variable S to zero, and then loops from 0 to 1200 assigning each time the value to a variable i, accumulating the sum of S + i in the variable S. **A final thing to notice is indentation.** In Python it is important to indent the code properly as this provides the right scope for variables (e.g. see that the line *S = S + 1* starts more to the right than the previous and following line -- this is because it is inside the for loop). You do not have to worry about this for the time being, we will get to this in a later practical...

How does this code work? How does the value of S and i change as the code is executed? These are questions that can be answered by the debugger.

To start the debugger, click on **Debug --> Start Debugging** (shortcut F5). The following small panel should pop up:

![](img/debug.png)

We will use it shortly, but before that, let's focus on what we want to track. On the left hand side of the main panel, a *Watch* panel appeared. This is where we need to add the things we want to monitor as the execution of the program goes. With respect to the code written above, we are interested in keeping an eye on the variables S, i and also of the expression S+i (that will give us the value of S of the next iteration). Add these three expressions in the watch panel (click on **+** to add new expressions). The watch panel should look like this:

![](img/watch.png)

do not worry about the message "*name X is not defined*", this is normal as no execution has taken place yet and the interpreter still does not know the value of these expressions.

The final thing before starting to debug is to set some breakpoints, places where the execution will stop so that we can check the value of the watched expressions. This can be done by hovering with the mouse on the left of the line number. A small reddish dot should appear, place the mouse over the correct line (e.g. the line corresponding to *S = S + 1* and click to add the breakpoint (a red dot should appear once you click).  

![](img/breakpoint.png)

Now we are ready to start debugging the code. Click on the green triangle on the small debug panel and you will see that the yellow arrow moved to the breakpoint and that the watch panel updated the value of all our expressions.

![](img/step0.png)

The value of all expressions is zero because the debugger stopped **before** executing the code specified at the breakpoint line (recall that S is initialized to 0 and that i will range from 0 to 1200). If you click again on the green arrow, execution will continue until the next breakpoint (we are in a for loop, so this will be again the same line - trust me for the time being).

![](img/step1.png)

Now i has been increased to 1, S is still 0 (remember that the execution stopped **before** executing the code at the breakpoint) and therefore S + i is now 1. Click one more time on the green arrow and values should update accordingly (i.e. S to 1, i to 2 and S + i to 3), another round of execution should update S to 3, i to 3 and S + i to 6. Got how this works? Variable i is increased by one each time, while S increases by i. You can go on for a few more iterations and see if this makes any sense to you, once you are done with debugging you can stop the execution by pressing the red square on the debug small panel. 


*Please take some more time to familiarize with Visual Studio Code (creating files, saving files, interacting with the debugger etc.) as in the next practicals we will take this ability for granted. Once you are done you can move on and do the following exercises.*

## Exercises

1. Compute the area of a triangle having base 120 units and height 33. Assign the result to a variable named area and print it.

2. Compute the area of a square having side S equal to 145 units. Assign the result to a variable named area and print it.

3. If you have not done so already, put the two previous scripts in two separate files (e.g. triangle_area.py and square_area.py and execute them from the terminal).

4. Write a small script (trapezoid.py) that computes the area of a trapezoid having major base (MB) equal to 30 units, minor base (mb) equal to 12 and height (H) equal to 17. Print the resulting area. Try executing the script from inside Visual Studio Code and from the terminal.

5. Rewrite the example of the sum of the first 1200 integers by using the following equation: $\sum\limits_{i=1}^n i = \frac{n (n+1)}{2}$



